services/src/test/java/io/grpc/protobuf/service/ProtoReflectionServiceTest.java

```diff
@@ -0,0 +1,312 @@
+import io.grpc.reflection.testing.ReflectionTestDepthTwoAlternateProto;
+import io.grpc.reflection.testing.ReflectionTestDepthTwoProto;
+import io.grpc.reflection.testing.ReflectionTestProto;
+import io.grpc.reflection.v1alpha.ExtensionRequest;
+import io.grpc.reflection.v1alpha.FileDescriptorResponse;
+import io.grpc.reflection.v1alpha.ServerReflectionRequest;
+import io.grpc.reflection.v1alpha.ServerReflectionResponse;
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.StreamRecorder;
+import io.grpc.util.MutableHandlerRegistry;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Tests for {@link ProtoReflectionService}.
+ */
+@RunWith(JUnit4.class)
+public class ProtoReflectionServiceTest {
+
+  private static final String TEST_HOST = "localhost";
+
+  private MutableHandlerRegistry handlerRegistry = new MutableHandlerRegistry();
+
+  private ProtoReflectionService reflectionService;
+
+  private ServerImpl server;
+
+  @Before
+  public void setUp() throws IOException {
+    reflectionService = new ProtoReflectionService();
+    server = InProcessServerBuilder.forName("proto-reflection-test")
+        .addService(reflectionService)
+        .addService(new ReflectableServiceGrpc.ReflectableServiceImplBase() {})
+        .fallbackHandlerRegistry(handlerRegistry)
+        .build();
+  }
+
+  @Test
+  public void listServices() throws Exception {
+    HashSet<ServiceResponse> originalServices = new HashSet<ServiceResponse>(
+        Arrays.asList(
+            ServiceResponse.newBuilder()
+                .setName("grpc.reflection.v1alpha.ServerReflection")
+                .build(),
+            ServiceResponse.newBuilder()
+                .setName("grpc.reflection.testing.ReflectableService")
+                .build())
+    );
+    assertServiceResponseEquals(originalServices);
+
+    ServerServiceDefinition dynamicService =
+        (new DynamicServiceGrpc.DynamicServiceImplBase() {}).bindService();
+    handlerRegistry.addService(dynamicService);
+    assertServiceResponseEquals(new HashSet<ServiceResponse>(
+        Arrays.asList(
+            ServiceResponse.newBuilder()
+                .setName("grpc.reflection.v1alpha.ServerReflection")
+                .build(),
+            ServiceResponse.newBuilder()
+                .setName("grpc.reflection.testing.ReflectableService")
+                .build(),
+            ServiceResponse.newBuilder()
+                .setName("grpc.reflection.testing.DynamicService")
+                .build())
+    ));
+
+    handlerRegistry.removeService(dynamicService);
+    assertServiceResponseEquals(originalServices);
+  }
+
+  @Test
+  public void fileByFilename() throws Exception {
+    final ServerReflectionRequest request =
```

Comment:
Don't use `final` for local variables.  It doesn't usually buy you much, and it adds boiler plate.